File: impute-rates.pl
#!/usr/bin/env perl

use v5.30;
use autodie;
use JSON::PP;

my %mha_assigns;
my %used_zip_codes;

my $INPUT_NAME  = $ARGV[0] or die usage();
my $OUTPUT_NAME = $ARGV[1] or die usage();
my $TOPOJSON_ZCTA_LAYER = "us_zcta500.geo"; # change if you change layer name in TopoJSON input

# Load BAH ZIP-code-to-MHA-assignment data
{
    open my $mha, '<', "sorted_zipmha23.txt";
    my @mhas = <$mha>;
    foreach my $mha_asgn (@mhas) {
        chomp($mha_asgn);
        my ($zip, $mha_id) = split(' ', $mha_asgn);
        $mha_assigns{$mha_id} //= 1;
        $used_zip_codes{$zip} = $mha_id;
    }
}

say scalar keys %mha_assigns, " MHAs, ", scalar keys %used_zip_codes, " ZIP codes in BAH database";

my $geoDB;

# Load the TopoJSON data with U.S. Census ZCTA geographic features. We will
# embed ZCTAs that match a BAH MHA area with the corresponding MHA ID.
{
    open my $topojson_file, '<', $INPUT_NAME;
    local $/; # slurp
    my $topojson_text = <$topojson_file>; # big!
    $geoDB = decode_json $topojson_text;
}

# The list of ZCTA geometries
my $obj_db = $geoDB->{objects}->{$TOPOJSON_ZCTA_LAYER}->{geometries};
my $upd_count = 0; # count imputations we make for comparison

say scalar @$obj_db, " objects in GEO DB";

for my $geometry (@$obj_db) {
    my $geoZCTA = $geometry->{properties}->{GEOID10}; # This is the 5-digit ZIP code as text
    my $mha = $used_zip_codes{$geoZCTA} // '';
    if ($mha) {
        $geometry->{properties}->{"DOD_BAH_MHA"} = $mha;
        $upd_count++;
        delete $used_zip_codes{$geoZCTA}; # any used ZIP codes not listed at end have no ZCTA match
    } else {
        say "GEO DB object $geoZCTA has no MHA match";
    }
}

say "Made $upd_count updates to the output GEO DB";
say scalar keys %used_zip_codes, " ZIP codes not used in any ZCTA entry";
my @non_fake_zips = grep { $used_zip_codes{$_} !~ /^ZZ/ } keys %used_zip_codes;
say "\t of these, ", scalar @non_fake_zips, " were not artificial MHAs";

# Write out the imputed data
{
    open my $topojson_file, '>', $OUTPUT_NAME;
    my $topojson_text = encode_json $geoDB;
    print $topojson_file $topojson_text;
}

say "\nUpdate written to $OUTPUT_NAME";

exit 0;

sub usage
{
    say <<~EOF;
        Encodes BAH data (sorted_zipmha23.txt must be available on disk)
        into the given input TopoJSON data containing U.S. Census Bureau ZCTAs.

        Output file is same as input, except that ZCTA geometry features
        with a ZIP code matching a DoD BAH MHA will have an additional property
        'DOD_BAH_MHA' containing the MHA ID for use in later topological merging

        $0 [input-file] [output-file]
        EOF

    return "Unable to proceed";
}

File: impute-rates.js
// Combine BAH data into the TopoJSON ZCTA data so that
// topomerge can combine like-MHA assignments into MHA
// geographic features.

const fs      = require('fs').promises;
const process = require('process');

// node, this script, input, output, obj-to-remove
if (process.argv.length != 4) {
    console.error(`Usage ${argv[1]} <input-file> <output-file>`);
    return 1;
}

const in_filename  = process.argv[2];
const out_filename = process.argv[3];
const TOPOJSON_ZCTA_LAYER = 'us_zcta'; // must match layer name in TopoJSON input

let zctaJson = ''; // main JSON data to fill in

async function loadMHAAssignmentData()
{
    const mha_lines = await fs.readFile("sorted_zipmha23.txt", 'utf-8');

    let mha_assigns    = new Set();
    let used_zip_codes = new Map;

    // ugly to read whole file in at once but the dataset isn't very large here
    for (const line of mha_lines.split(/\r?\n/)) {
        const [zip, mha_id] = line.split(' ');
        if (!zip || !mha_id) {
            throw new Error(`Undefined ZIP or MHA entry ${line}`);
        }

        mha_assigns.add(mha_id);
        used_zip_codes.set(zip, mha_id);
    }

    return [mha_assigns, used_zip_codes];
}

async function loadZCTAData(zctaFile)
{
    const zctaBuf = await fs.readFile(zctaFile);

    return JSON.parse(zctaBuf);
}

function buildAugmentedZCTAData(zctaJson, mhas, zips)
{
    let zctaUpdateCount = 0;
    const zctaTotal = zctaJson.objects[TOPOJSON_ZCTA_LAYER].geometries.length;

    for (let zcta of zctaJson.objects[TOPOJSON_ZCTA_LAYER].geometries) {
        const zcta_zip = zcta.properties.GEOID20; // 5-digit ZIP code, name changes in datasets sometimes

        if (zips.has(zcta_zip)) {
            zcta.properties.DOD_BAH_MHA = zips.get(zcta_zip);
            zctaUpdateCount++;
        } else {
            zcta.properties.DOD_BAH_MHA = 'unused';
        }
    }

    console.log(`${zctaUpdateCount} / ${zctaTotal} ZCTAs assigned to an MHA`);
    console.log(`${zctaUpdateCount} / ${zips.size} MHA ZIP codes assigned to a ZCTA`);

    if (zctaUpdateCount == 0) {
        throw new Error("Something must have gone wrong, aborting.");
    }

    return zctaJson;
}

const mhaPromise  = loadMHAAssignmentData();
const zctaPromise = loadZCTAData(in_filename);

Promise.all([mhaPromise, zctaPromise])
    .then(([mhaResults, zctaResults]) => {
        const [mhas, zips] = mhaResults;

        console.log(`${mhas.size} MHAs defined across ${zips.size} ZIP codes`);
        console.log(`${zctaResults.objects.us_zcta.geometries.length} ZCTAs loaded`);

        const mergedZcta = buildAugmentedZCTAData(zctaResults, mhas, zips);

        // write results
        return fs.writeFile(out_filename, JSON.stringify(mergedZcta));
    }).catch((err) => {
        console.warn(`Caught error ${err}`);
        process.exit(1);
    });
